src/output/gtk-shell.cpp

```diff
@@ -2,75 +2,227 @@
+{
+    auto surface = static_cast<wl_resource*> (wl_resource_get_user_data(resource));
+    wayfire_view view = wf::wl_surface_to_wayfire_view(surface);
+    if (view && view != nullptr)
+        view->erase_data("gtk-shell-modal");
+}
+
+
+/**
+ * The surface requests focus, for example single instance applications like
+ * gnome-control-center, gnome-clocks, dconf-editor are single instance and if
+ * they are already running and launched again, this will request that they get focused.
+ * This function is superseeded by handle_gtk_surface_request_focus a newer equivalelent
+ * used by gtk-applications now. This function is for compatibility reasons.
+ */
+static void handle_gtk_surface_present(wl_client *client, wl_resource *resource, uint32_t time)
+{
+    auto surface = static_cast<wl_resource*> (wl_resource_get_user_data(resource));
+    wayfire_view view = wf::wl_surface_to_wayfire_view(surface);
+    if (view && view != nullptr)
+    {
+        view_self_request_focus_signal data;
+        data.view = view;
+        view->get_output()->emit_signal("view-self-request-focus", &data);
+    }
+}
+
+
+/**
+ * The surface requests focus, for example single instance applications like
+ * gnome-control-center, gnome-clocks, dconf-editor are single instance and if
+ * they are already running and launched again, this will request that they get focused.
+ */
+static void handle_gtk_surface_request_focus(struct wl_client *client, struct wl_resource *resource, const char* startup_id)
+{
+    auto surface = static_cast<wl_resource*> (wl_resource_get_user_data(resource));
+    wayfire_view view = wf::wl_surface_to_wayfire_view(surface);
+    if (view && view != nullptr)
+    {
+        view_self_request_focus_signal data;
+        data.view = view;
+        view->get_output()->emit_signal("view-self-request-focus", &data);
+    }
+}
+
+
+/**
+ * Destroys the gtk_surface object.
+ * TODO: May this be required for some kind of leak?
+ */
+static void handle_gtk_surface_destroy(wl_resource *resource) {}
+
+
+/**
+ * Supported functions of the gtk_surface_interface implementation
+ */
 const struct gtk_surface1_interface gtk_surface1_impl = {
-	.set_dbus_properties = handle_gtk_surface_set_dbus_properties,
-	.set_modal = handle_gtk_surface_set_modal,
-	.unset_modal = handle_gtk_surface_unset_modal,
-	.present = handle_gtk_surface_present,
+    .set_dbus_properties = handle_gtk_surface_set_dbus_properties,
+    .set_modal = handle_gtk_surface_set_modal,
+    .unset_modal = handle_gtk_surface_unset_modal,
+    .present = handle_gtk_surface_present,
+    .request_focus = handle_gtk_surface_request_focus,
 };
 
-static void handle_gtk_shell_get_gtk_surface(wl_client *client, wl_resource *resource,
-    uint32_t id, wl_resource *surface)
+/**
+ * Passes the gtk_surface object to the client.
+ */
+static void handle_gtk_shell_get_gtk_surface(wl_client *client,
+    wl_resource *resource,
```

Comment:
I usually try to binpack arguments on the same line, as it was before ...

Reply:
done.